Common models

ngurenyaga opened the pull request:
This pull request establishes the core of the MFL 2.0 API.

Commit:
Add common services to thier own app.

Commit:
Add stubs for facility models.

Commit:
Models checkpoint.

Commit:
Add AUTH_USER_MODEL to settings.

Commit:
Checkpoint

Commit:
Fix custom user issues

Commit:
Add migrations

Commit:
Checkpoint

Commit:
Merge branch 'master' of github.com:MasterFacilityList/mfl_api into facility

Commit:
Add tests for the contact model.

Commit:
Add tests for county, province ans constituency model.

Commit:
Add tests for sublocation model.

Commit:
Test sub locations __unicode__

Commit:
Fix pep 8 issues in the common app

Commit:
Checkpoint

Commit:
Add tests for role models

Commit:
Checkpoint:
   1. Naive RBAC
   2: facitilities api

Commit:
Add cors

ngurenyaga commented on .coveragerc:
```diff
@@ -1,6 +1,6 @@
 [run]
 branch=True
-source=minion
+source=minion, common
```
Now that you already have a real Django app, do you still need to have the `minion` app? I'd recommend that you remove it.

ngurenyaga commented on common/admin.py:
```diff
@@ -0,0 +1,14 @@
+from django.contrib import admin
```
Remove the admin, do proper data loading

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,163 @@
+LOGGER = logging.getLogger(__file__)
+
+
+class AbstractBase(models.Model):
```
Missing docstring; what is this, why is it there?

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,163 @@
+        abstract = True
+
+
+class RegionAbstractBase(AbstractBase):
```
Missing docstring; what is this, why is it there?

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,163 @@
+        super(Contact, self).save(*args, **kwargs)
+
+
+class Province(RegionAbstractBase):
```
Provinces are now obsolete as geographic units.

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,163 @@
+        abstract = True
+
+
+class Contact(AbstractBase):
```
Change this to be a smaller contact table; just `contact_type` and `contact`; the usage pattern -> anything that can have contacts e.g a Facility will have a M2M relationship with this; it can have zero or more contacts.

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,163 @@
+    landline = models.CharField(max_length=100)
+    mobile = models.CharField(max_length=10)
+
+    def __unicode__(self):
```
This will obviously change when the `Contact` model is rewritten.

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,163 @@
+
+
+class County(RegionAbstractBase):
+    Province = models.ForeignKey(Province, null=True, blank=True)
```
No more provinces.

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,163 @@
+    def __unicode__(self):
+        return self.name
+
+    class Meta:
```
No point having an empty `Meta`

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,163 @@
+        return self.name
+
+
+class District(RegionAbstractBase):
```
No more districts; for now, just remove this.

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,163 @@
+        return self.name
+
+
+class Division(RegionAbstractBase):
```
Remove this

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,163 @@
+        return self.name
+
+
+class Location(RegionAbstractBase):
```
Remove this for now.

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,163 @@
+        return self.name
+
+
+class SubLocation(RegionAbstractBase):
```
Remove this for now.

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,163 @@
+        return self.division.constituency
+
+
+class Feedback(models.Model):
```
The feedback model is a luxury; remove this for now.

ngurenyaga commented on config/__init__.py:
```diff
@@ -0,0 +1,20 @@
+from django.utils.version import get_version
+
+VERSION = (1, 8, 0, 'alpha', 0)
```
I don't understand this 1.8.0alpha business; not to mention that Django 1.8 is now officially released.

ngurenyaga commented on config/settings/__init__.py:
```diff
@@ -1 +1,20 @@
-from .base import *  # noqa
+# from django.utils.version import get_version
```
Code scars; remove the commented lines; Also, why are they doing the same thing done in `config/__init__.py`?

ngurenyaga commented on config/settings/base.py:
```diff
@@ -67,3 +76,40 @@
+    'DEFAULT_MODEL_SERIALIZER_CLASS':
+        'rest_framework.serializers.ModelSerializer',
+    'DEFAULT_PERMISSION_CLASSES': (
+        # 'roles.permissions.MFLPermission',
```
Let the default be `IsAuthenticated`. That is a saner default.

ngurenyaga replied:
Remove the `AllowAny`

ngurenyaga commented on config/settings/base.py:
```diff
@@ -67,3 +76,40 @@
+        'rest_framework.filters.DjangoFilterBackend',
+    ),
+    'DEFAULT_PARSER_CLASSES': (
+        'djangorestframework_camel_case.parser.CamelCaseJSONParser',
```
I am ambivalent about the use of camel case; since the front-end has not started, let us leave the parsers and renderers at the default "underscoreized" JSON format.

Also:
- add in XML parser / renderer
- add a CSV renderer: https://github.com/mjumbewu/django-rest-framework-csv
- add a Pandas ( Excel ) renderer: https://github.com/wq/django-rest-pandas

The first one is for compatibility purposes; the last two will be the primary mechanism for export of data ( export will be via a custom DRF serializer, with all applicable DRF filters being usable ).

ngurenyaga commented on fabfile.py:
```diff
@@ -15,3 +19,26 @@ def test():
+def setup(*args, **kwargs):
+    no_sudo = True if 'no-sudo' in args else False
+    bootstrap_sql = kwargs['sql'] if 'sql' in kwargs else None
+    psql("DROP DATABASE IF EXISTS mfl", no_sudo)
```
For this section, it is more newbie friendly and DRY if you import `django.conf.settings` ( that will require you to add a few lines of Django configuration on top of the fabfile ) and use string interpolation to interpolate in the default database name, database username and database password. 

That will ensure that there is only one place to change database settings.

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,58 @@
+    ('NOT_OPERATIONAL', 'The facility is not operating'),
+)
+
+FACILITY_TYPES = (
```
This should be a model, with default, data.

One of the big failings of MFL 1.0 was the lack of configurability of changeable choices like this.

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,58 @@
+from django.db import models
+from common.models import AbstractBase, SubLocation
+
+FACILLITY_STATUS = (
```
This should be a simple model, with default data.

For this project, apply a simple rule: NO CHOICE FIELDS, UNLESS PREVIOUSLY DISCUSSED AND AGREED.

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,58 @@
+class Owner(AbstractBase):
+    name = models.CharField(max_length=100, unique=True)
+    description = models.TextField(null=True, blank=True)
+    code = models.CharField(max_length=100)
```
Let the `code` too be unique.

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,58 @@
+class Service(AbstractBase):
+    name = models.CharField(max_length=255)
+    description = models.TextField(null=True, blank=True)
+    code = models.CharField(max_length=100, null=True, blank=True)
```
Let the `code` be non nullable. In the event that a null value is supplied, supply a callable that can be evaluated to create a default `code` by applying a transformation to the `name`

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,58 @@
+class Facility(AbstractBase):
+    name = models.CharField(max_length=100, unique=True)
+    code = models.CharField(max_length=100, unique=True)
+    latitude = models.CharField(max_length=255)
```
I'd prefer to have the GIS stuff abstracted into a separate models e.g `FacilityCoordinates` - with as `OneToOneField` relation back to the `Facility`.

There will be more "adventures" with the co-ordinate data than with the facilities data.

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,58 @@
+    open_whole_day = models.BooleanField(default=False)
+    open_whole_week = models.BooleanField(default=False)
+    status = models.CharField(
+        max_length=50, choices=FACILLITY_STATUS)
```
Spelling error in choices; not relevant anyway - the choices will be replaced.

ngurenyaga commented on facilities/serializers.py:
```diff
@@ -0,0 +1,52 @@
+        AbstractFieldsMixin, serializers.ModelSerializer):
+    services = ServiceSerializer(many=True, required=False)
+
+    def create(self, validated_data):
```
This is good code; but it will be simpler if we keep the API to simple, reliable CRUD; @liyosi and @owagaantony can create front-end wrappers that do this stuff that you are doing here.

i.e. unless there is a direct request and a good reason, let M2M fields be manipulated through the join tables.

For the front-enders: @owagaantony @sombo09 @liyosi  - design the data input forms so that _it is natural to save the "header" information first_ then save the other items e.g linked services one at a time.

@bmarika - Remove this crutch so that the FE team can be forced to do thing right from the word go.

ngurenyaga replied:
You mis-understood my intention in the previous commit; remove this `create` method; the front-enders will use APIs on the join tables directly.

ngurenyaga commented on facilities/views.py:
```diff
@@ -0,0 +1,43 @@
+class OwnerListView(generics.ListCreateAPIView):
+    queryset = Owner.objects.all()
+    serializer_class = OwnerSerializer
+    filter_fields = ()
```
The models are few enough that we can do filter classes for everything.

At first blush, that appears to be overkill; it is not ( will explain later ).

ngurenyaga commented on facilities/views.py:
```diff
@@ -0,0 +1,43 @@
+    lookup_field = 'id'
+
+
+class ServiceListView(generics.ListCreateAPIView):
```
Also - let every list have ordering configured: http://www.django-rest-framework.org/api-guide/filtering/#orderingfilter

ngurenyaga commented on users/models.py:
```diff
@@ -0,0 +1,98 @@
+    is_incharge = models.BooleanField(default=False)
+    contact = models.ForeignKey(Contact, null=True, blank=True)
+    county = models.ForeignKey(County, null=True, blank=True)
+    is_national = models.BooleanField(default=False)
```
These boolean fields are a noble effort, but I feel that they have not been thought through completely;

Some questions:
- why is there a `county` FK in the user model, yet there is a join table that links users to counties?
- what happens if a user leaves a county?
- what about sub-counties? [ maybe constituencies ]? That will be the first point of data input.
- why do we have an `is_incharge` field, yet that information is held in the join table that links users to counties?
- can a user have more than one contact? [ esp. after refactoring the `Contact` model ]
- how does all this interact with filtering in the serializers / views? Do we need a custom permissions backend? What would be the behavior of that custom permissions backend?

@ngash - Consulting needed here.

bmarika replied:
Let me bring in a minimal user model without RBAC and contacts and counties. This will be handled in subsequent PR

ngurenyaga commented on users/models.py:
```diff
@@ -0,0 +1,98 @@
+        super(MflUser, self).save(*args, **kwargs)
+
+
+class InchargeCounties(AbstractBase):
```
The name of this model could use some improvement.

ngurenyaga commented:
@bmarika A first review checklist, to help you get this foundation merged in good form:
- [x] Remove the `minion` app
- [x] Remove every `admin.py`
- [x] Port in the data loader ( copy the current / latest app version here and install it; no pip )
- [x] Add in docstrings for `AbstractBase` and `RegionAbstractBase` ( and every other abstract base you add ); those are important enough to need an explanation
- [x] Remove obsolete geographic units e.g provinces
- [x] Refactor the contacts model
- [x] Remove the feedback model; we focus on the core data model for now
- [x] Let the default REST_FRAMEWORK permission be `IsAuthenticated`; completely remove `AllowAny`
- [x] Remove the camel case parsers, add in XML parsing / rendering and CSV & Pandas _rendering_
- [x] `config/__init__.py` and `confing/settings/__init__.py` need fixing ( or removing - you did not go the full hog with versioning, you can live to fight that battle another day )
- [x] DRYness in the fabfile; should interpolate database settings from `django.conf.settings`
- [x] Remove the empty `Meta` in the `County` model ( or add stuff that matters to it, if that was the intention )
- [x] Facility status and facility types should be models; NO CHOICE FIELDS; let the models have default data
- [x] Make `code` unique in `Owner` **and** implement a callable that can reliably ( no race conditions ) auto-generate codes if one is not supplied at create time
- [x] Make `code` in `Services` unique and non-nullable; implement a callable to generate sensible default codes if not supplied
- [x] Move facility GIS co-ordinate stuff into a separate model
- [x] Remove the "special" facility serializer that attempts to handle POSTed facility payloads that have services inlined
- [x] Filter classes for all list endpoints; not negotiable
- [x] Ordering for every list endpoint
- [x] Sort out boolean fields in user model ( see comments )
- [x] A better name for the `InChargeCounties` models
- [x] Add a CircleCI embeddable status badge to the README: https://circleci.com/docs/status-badges
- [x] Review the requirements and freshen them e.g Django 1.8 is now final; I strongly suggest that yu use `pip-review` from https://github.com/nvie/pip-tools
- [x] Write a `README` that gives usable starter instructions **and has a proper heading hierarchy**
- [x] Complete test coverage

After this, multiple developers can now work on individual tickets ( much smaller PRs ).

There will be a second review; more detailed than the first.

Commit:
Review checkpoint.

Commit:
Updates to readme

Commit:
Minor read me fix

Commit:
Add information on data formats.

Commit:
Fix typo in data boostrap name

Commit:
Add data_bootstrap app to installed apps

Commit:
Move requirements to setup.py.

Commit:
Fix formatting.

Commit:
Change circleci status badge style to shield.

Commit:
Add  link to ehealth.

Commit:
Update typography.

Commit:
Update typography.

Commit:
Remove model_mommy from requirements

Commit:
Fix pep 8

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,83 @@
+    owner = models.ForeignKey(Owner)
+
+    def generate_code(self):
+        random_number = random.randint(10000, 1000000)
```
The standards call for a _sequential_ facility code ( "assign the next unassigned number" ). In order for this to be reliable, it needs to occur via a _database level sequence_.

Also - as the creator of a version 2 API - you will have to account for all the existing codes when initializing that database sequence AND preserve the numbering scheme that was pioneered in version 1.

Kindly consult me offline on how to go about this.

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,83 @@
+
+class Facility(AbstractBase):
+    name = models.CharField(max_length=100, unique=True)
+    code = models.CharField(max_length=100, unique=True)
```
( see the comment on the `generate_code` function first )

It will be easier - even on testing - if you make `generate_code` a module level no-args callable that is then supplied as the `default` for the `code` field.

In order to distinguish it from other code generation callables, give it a descriptive name e.g `generate_sequential_facility_code`. I suspect that you will have other code generating functions, and that opportunities for refactoring / code sharing will be seen without too much effort.

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,107 @@
+    Model to hold the common attributes of a region
+    """
+    name = models.CharField(max_length=100, unique=True)
+    code = models.CharField(max_length=100, unique=True)
```
See comments made under `Facility` about having a default code generating function.

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,83 @@
+class Owner(AbstractBase):
+    name = models.CharField(max_length=100, unique=True)
+    description = models.TextField(null=True, blank=True)
+    code = models.CharField(max_length=100, unique=True)
```
See comments made under `Facility` about having a default code generating function.

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,83 @@
+    def generate_code(self):
+        return "{}{}".format("OWNER", str(self.id))
+
+    def save(self, *args, **kwargs):
```
If you move to have a default code generating function ( `default` keyword in field definition ), you will not need this custom `save` method.

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,83 @@
+class Service(AbstractBase):
+    name = models.CharField(max_length=255)
+    description = models.TextField(null=True, blank=True)
+    code = models.CharField(max_length=100, unique=True)
```
See comments made under `Facility` about having a default code generation function.

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,83 @@
+    def __unicode__(self):
+        return self.name
+
+    def generate_code(self):
```
This 'approach' to generation of codes will obviously not work...because IDs are _assigned through a database sequence_.

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,83 @@
+class Facility(AbstractBase):
+    name = models.CharField(max_length=100, unique=True)
+    code = models.CharField(max_length=100, unique=True)
+    gis = models.OneToOneField(GIS, null=True, blank=True)
```
I'd rather have the `GIS` mode have the `OneToOneField` relationship with the `Facility` ( inverse of what you have ).

Then - when you move this field, make it non-nullable.

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,83 @@
+    def generate_code(self):
+        return "{}{}".format(self.name, str(self.id))
+
+    def save(self, *args, **kwargs):
```
You will not require a custom `save` method when you move to a `default=` callable.

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,83 @@
+        super(Service, self).save(*args, **kwargs)
+
+
+class GIS(AbstractBase):
```
This model could use a rename...something like `FacilityGIS` ( see a different comment made above ).

For this project, the only "things" that will have GIS co-ordinates are facilities.

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,83 @@
+class GIS(AbstractBase):
+    latitude = models.CharField(max_length=255)
+    longitude = models.CharField(max_length=255)
+    is_classified = models.BooleanField(default=False)
```
Your fields need pervasive `help_text`...especially fields like this one.

ngurenyaga commented on README.rst:
```diff
@@ -1 +1,86 @@
+Data Formats Supported
+----------------------
+The formats supported are xml, json, csv and xlsx.
+The default data format is json. To access xml append ``?format=xml`` to an endpoint. To access data in csv append ``?format=csv`` and similarly to access xlsx append ``?format=xlsx``. 
```
There is another way to do content negotiation...through headers.

Document that too.

ngurenyaga commented on README.rst:
```diff
@@ -1 +1,86 @@
+    :target: https://circleci.com/gh/MasterFacilityList/mfl_api
+
+This is the backend for the Master Facility list. 
+It provides an api through which one is able to access various resources provided by master facility list; Including facilities and their services.
```
Grammar; "API" is usually capitalized; "Master Facility List" is a compound noun - use sentence case. The first "I" after the semicolon should be lower-case ( that is not a sentence boundary ).

ngurenyaga commented on README.rst:
```diff
@@ -1 +1,86 @@
+
+    * Run migrations.
+
+    * Load demo data if the project is in development mode.
```
What does "the project is in development mode" mean? Does it mean that `DEBUG` is `True` in settings?

[ clarify ]

ngurenyaga commented on README.rst:
```diff
@@ -1 +1,86 @@
+
+2. Activate the virtualenv and run ``python setup.py install`` while in the project folder.
+
+Explore the api
```
"API" should be capitalized

ngurenyaga commented on README.rst:
```diff
@@ -1 +1,86 @@
+
+Explore the api
+---------------
+Once all the steps above have been carried out, the system is ready to run and one can explore the api. 
```
Capitalize "API"

ngurenyaga commented on README.rst:
```diff
@@ -1 +1,86 @@
+---------------
+Once all the steps above have been carried out, the system is ready to run and one can explore the api. 
+
+There are two ways to explore the api:
```
Capitalize "API"

ngurenyaga commented on README.rst:
```diff
@@ -1 +1,86 @@
+
+There are two ways to explore the api:
+
+**a. Through swagger**
```
"Swagger" is a proper noun ( should start with a capital letter ).

You'd do well to also link to http://swagger.io/ - provide background for those that do not know what Swagger is.

ngurenyaga commented on README.rst:
```diff
@@ -1 +1,86 @@
+**a. Through swagger**
+   This is the recommended way as one can easily try  ``GET``, ``POST``, ``PUT``,  ``PATCH``, ``DELETE``, ``OPTIONS`` on an endpoint if they are allowed.
+
+   To explore the endpoint throught swagger visit the url ``api/explore/``. 
```
Capitalize the proper noun "Swagger".

ngurenyaga commented on README.rst:
```diff
@@ -1 +1,86 @@
+
+**b. Through accessing the urls directly**
+
+Below is a listing of the major urls in the project:
```
It would be much better if you created and maintained an API root.

The old technique:

``` python
from rest_framework.decorators import api_view
from rest_framework.response import Response
from rest_framework.reverse import reverse


@api_view(('GET',))
def api_root(request, format=None):
    return Response({
        'users': reverse('user-list', request=request, format=format),
        'snippets': reverse('snippet-list', request=request, format=format)
    })
```

Newer approaches might be possible using: http://www.django-rest-framework.org/api-guide/metadata/

Either way - let there be an API root, then the `README` only needs to point new users to the API root.

ngurenyaga commented on README.rst:
```diff
@@ -1 +1,86 @@
+``api/facilities/``
+
+**Services**
+``api/facilities/services/``
```
The `README` on access to the API conveniently avoids any mention of authentication.

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,107 @@
+
+LOGGER = logging.getLogger(__file__)
+
+CONTACT_TYPES = (
```
Let contact types too be configurable via a simple model ( only a unique `name` field ).

Yes, it appears to be overkill at present; do it still.

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,107 @@
+    It will provide audit fields that will keep track of when a model
+    is created or updated and by who.
+    """
+
```
It is good that every model descends from `AbstractBase` ( enforce that rule firmly ).

Remember why we moved to Django 1.8 even before it was released? Because we wanted **UUID primary keys** - see https://docs.djangoproject.com/en/1.8/ref/models/fields/#django.db.models.UUIDField 

( add an explicit `id` field ).

These fields will make federation much easier when the time for that comes.

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,107 @@
+    ('EMAIL', 'email'),
+    ('LANDLINE', 'landline'),
+    ('MOBILE', 'Mobile Phone number'),
+
```
Spurious space.

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,107 @@
+
+class RegionAbstractBase(AbstractBase):
+    """
+    Model to hold the common attributes of a region
```
This docstring does not tell us what a "region" is.

Remember - a docstring's job is not to explain things that are already manifest in the code, but to explain the _why_ and expose any potential traps / pitfalls.

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,107 @@
+    contact_type = models.CharField(choices=CONTACT_TYPES, max_length=100)
+
+    def __unicode__(self):
+        return str(self.id)
```
Two things...

_First_, a `__unicode__` that simply returns the `id` is not actually useful. It will become significantly less useful when we move to UUID PKs. Each `__unicode__` needs to be hand crafted for its model.

_Second_ - you've opened a Pandora's box by adding `__unicode__` methods to some models; now _all models will need to have them_

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,107 @@
+    plot_number = models.CharField(max_length=100)
+
+    def __unicode__(self):
+        return str(self.id)
```
See previous comments about `__unicode__`

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,107 @@
+
+class County(RegionAbstractBase):
+    def __unicode__(self):
+        return self.name
```
If the `RegionAbstractBase` has a working `__unicode__` then you will not need this.

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,107 @@
+    county = models.ForeignKey(County)
+
+    def __unicode__(self):
+        return self.name
```
See previous comments about `__unicode__`

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,107 @@
+    county = models.ForeignKey(County)
+
+    def __unicode__(self):
+        return self.name
```
See previous comments about `__unicode__`

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,107 @@
+        return str(self.id)
+
+
+class County(RegionAbstractBase):
```
No docstring; particularly important here.

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,107 @@
+        return self.name
+
+
+class SubCounty(RegionAbstractBase):
```
A good docstring is needed here.

ngurenyaga commented on common/models.py:
```diff
@@ -0,0 +1,107 @@
+        return self.name
+
+
+class Constituency(RegionAbstractBase):
```
A good docstring is needed here.

ngurenyaga commented on common/serializers.py:
```diff
@@ -0,0 +1,55 @@
+    """
+    Injects the fields in the abstract base model as a model
+    instance is being saved.
+
```
Spurious space.

ngurenyaga commented on common/views.py:
```diff
@@ -0,0 +1,73 @@
+    ConstituencySerializer)
+
+
+class FilterViewMixin(object):
```
Good start, but this is still a blunt instrument. 

It will be more maintainable if done via a custom permissions class [ http://www.django-rest-framework.org/api-guide/permissions/#custom-permissions ] that can be added to the views that need it ( while other views are left with the default ). The internal implementation of that custom permissions class will be very similar to the code you have here.

The code you have below is also "too branchy". It could be rewritten in a shorter way as follows:

``` python
def get_queryset(self):
    user = self.request.user
    if isinstance(user, AnonymousUser) or user.is_national:
        return self.queryset  # Leave the queryset unbothered
    elif user.is_incharge:
       return self.queryset.filter(county=user.county)  # Blind spot; what about sub-counties?
    else:
        return [] # Isn't it better to raise a custom exception here, rather than fail silently. This should not happen with a properly set up account
```

The rewritten sketch above makes it easy to follow the logic for each of the three possible "exit points".

ngurenyaga commented on common/views.py:
```diff
@@ -0,0 +1,73 @@
+class ContactDetailView(generics.RetrieveUpdateDestroyAPIView):
+    queryset = Contact.objects.all()
+    serializer_class = ContactSerializer
+    lookup_field = 'id'
```
These `lookup_fields` do not change the default behavior ( lookup by `pk`, which is itself a Django alias of `id` - unless you have a funky model setup ).

I suggest that they all be removed.

ngurenyaga commented on common/views.py:
```diff
@@ -0,0 +1,73 @@
+
+class ContactView(generics.ListCreateAPIView):
+    queryset = Contact.objects.all()
+    serializer_class = ContactSerializer
```
You ticked the Review 1 checklist that asserted that "every list view must have ordering fields". That "contract" was clearly not followed through to the end. Kindly review _all_ the list views.

ngurenyaga commented on common/views.py:
```diff
@@ -0,0 +1,73 @@
+class CountyDetailView(generics.RetrieveUpdateDestroyAPIView):
+    queryset = County.objects.all()
+    serializer_class = CountySerializer
+    lookup_field = 'id'
```
See previous comments about `LookupField`

ngurenyaga commented on common/views.py:
```diff
@@ -0,0 +1,73 @@
+class SubCountyDetailView(generics.RetrieveUpdateDestroyAPIView):
+    queryset = Constituency.objects.all()
+    serializer_class = SubCountySerializer
+    lookup_field = 'id'
```
See previous comments about `LookupField`

ngurenyaga commented on common/views.py:
```diff
@@ -0,0 +1,73 @@
+class ConstituentcyDetailView(generics.RetrieveUpdateDestroyAPIView):
+    queryset = Constituency.objects.all()
+    serializer_class = ConstituencySerializer
+    lookup_field = 'id'
```
See previous comments about `LookupField`

Commit:
Add tests for common views.

Commit:
Add dist and build to gitignore.

ngurenyaga commented on common/views.py:
```diff
@@ -0,0 +1,73 @@
+class ConstituencyView(generics.ListCreateAPIView):
+    queryset = Constituency.objects.all()
+    serializer_class = ConstituencySerializer
+    filter_fields = ('county', )
```
Another "contract violation" - you ticked the assertion about "Filter Classes" in the first review checklist.

ngurenyaga commented on config/settings/base.py:
```diff
@@ -18,20 +18,30 @@
+CORS_ORIGIN_ALLOW_ALL = True
+
+AUTH_USER_MODEL = 'users.MflUser'
 MIDDLEWARE_CLASSES = (
```
Add a space between `AUTH_USER_MODEL` and `MIDDLEWARE_CLASSES`

ngurenyaga commented on config/settings/base.py:
```diff
@@ -67,3 +77,39 @@
+REST_FRAMEWORK = {
+    # Use hyperlinked styles by default.
+    'DEFAULT_MODEL_SERIALIZER_CLASS':
+        'rest_framework.serializers.ModelSerializer',
```
The comment `# Use hyperlinked styles by default.` does not agree with the actual default that you have configured ( `'rest_framework.serializers.ModelSerializer'` ).

ngurenyaga commented on data_bootstrap/management/commands/bootstrap.py:
```diff
@@ -0,0 +1,20 @@
+                # check if it's a glob
+                for filename in glob.glob(suggestion):
+                    process_json_file(filename)
+        print "Done loading"
```
Do not use `print` in management commands. I'll let the following excerpt from the Django docs speak for itself:

```
Note

When you are using management commands and wish to provide console output, you should write to self.stdout and self.stderr, instead of printing to stdout and stderr directly. By using these proxies, it becomes much easier to test your custom command. Note also that you don’t need to end messages with a newline character, it will be added automatically, unless you specify the ending parameter:

self.stdout.write("Unterminated line", ending='')
```

PS: Backport that change to the source of this library.

ngurenyaga commented on fabfile.py:
```diff
@@ -15,3 +19,30 @@ def test():
+
+    psql("DROP DATABASE IF EXISTS {}".format(db_name), no_sudo)
+    psql("DROP USER IF EXISTS {}".format(db_user), no_sudo)
+    psql("CREATE USER mfl WITH SUPERUSER CREATEDB "
```
The `mfl` user is still hard coded.

ngurenyaga commented on fabfile.py:
```diff
@@ -15,3 +19,30 @@ def test():
+    psql("DROP USER IF EXISTS {}".format(db_user), no_sudo)
+    psql("CREATE USER mfl WITH SUPERUSER CREATEDB "
+         "CREATEROLE LOGIN PASSWORD '{}'".format(db_pass), no_sudo)
+    psql('CREATE DATABASE mfl', no_sudo)
```
The `mfl` database name is still hard coded - should be interpolated from settings.

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,83 @@
+    code = models.CharField(max_length=100, unique=True)
+
+    def generate_code(self):
+        return "{}{}".format("OWNER", str(self.id))
```
See previous comments about why this style of generating codes will not work

( TL;DR - at the point of `save`, we do not know the `id` yet; plus - `id`s will soon be UUIDs ).

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,83 @@
+
+
+class FacilityStatus(AbstractBase):
+    name = models.CharField(max_length=100)
```
Make the `name` unique

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,83 @@
+
+
+class FacilityType(AbstractBase):
+    name = models.CharField(max_length=100)
```
Make the `name` unique

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,83 @@
+    services = models.ManyToManyField(Service)
+    number_of_beds = models.PositiveIntegerField(default=0)
+    number_of_cots = models.PositiveIntegerField(default=0)
+    open_whole_day = models.BooleanField(default=False)
```
Look at better ways to model opening hours and opening days; `open_whole_day` and `open_whole_week` will not cut it.

The days of the week are known...
The number of hours in the day is known...

It should be possible to add a subsidiary model that holds this information in a better way.

For backwards compatibility and to provide API syntactic sugar, `open_whole_day` and `open_whole_week` can come back as `@property`s and `@property.setter`s.

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,83 @@
+    class Meta:
+        verbose_name_plural = 'Facilities'
+
+    def save(self, *args, **kwargs):
```
This should go away as soon as you move to using a code generating callable as the default for the `code` field.

ngurenyaga commented on facilities/views.py:
```diff
@@ -0,0 +1,42 @@
+class OwnerDetailView(generics.RetrieveUpdateDestroyAPIView):
+    queryset = Owner.objects.all()
+    serializer_class = OwnerSerializer
+    lookup_field = 'id'
```
All the lookup fields that do not change default behavior need to be removed.

ngurenyaga commented on config/settings/base.py:
```diff
@@ -67,3 +77,39 @@
+        'roles.permissions.MFLPermission',
+    ),
+    'DEFAULT_FILTER_BACKENDS': (
+        'rest_framework.filters.DjangoFilterBackend',
```
Add the `OrderingFilter` backend here too...there was a comment to that effect in the first review.

See: http://www.django-rest-framework.org/api-guide/filtering/#orderingfilter

ngurenyaga commented on facilities/views.py:
```diff
@@ -0,0 +1,42 @@
+class ServiceDetailView(generics.RetrieveUpdateDestroyAPIView):
+    queryset = Service.objects.all()
+    serializer_class = ServiceSerializer
+    lookup_field = 'id'
```
Remove all `lookup_field`s that do not change default behavior.

ngurenyaga commented on facilities/views.py:
```diff
@@ -0,0 +1,42 @@
+class FacilityListView(generics.ListCreateAPIView):
+    queryset = Facility.objects.all()
+    serializer_class = FacilitySerializer
+    filter_fields = ('owner',)
```
See previous comments about using filter classes.

ngurenyaga commented on mfl.sql:
```diff
@@ -0,0 +1,3376 @@
+--
```
What is the purpose of this `mfl.sql` file dump? Remove it.

ngurenyaga commented on facilities/views.py:
```diff
@@ -0,0 +1,42 @@
+class FaciltyDetailView(generics.RetrieveUpdateDestroyAPIView):
+    queryset = Facility.objects.all()
+    serializer_class = FacilitySerializer
+    lookup_field = 'id'
```
Remove `lookup_field`s that do not change default behaviour.

ngurenyaga commented on roles/models.py:
```diff
@@ -0,0 +1,45 @@
+from django.db import models
```
Looking at the entirety of this file...why not use the models in `django.contrib.auth` [ https://docs.djangoproject.com/en/1.8/ref/contrib/auth/ ].

Even for the custom user model, the closer you stick to the abstract bases / mixins Django provides, the better.

In my view, this file consists of some entirely unnecessary re-invention - or, if necessary, not completely thought through.

ngurenyaga commented on roles/permissions.py:
```diff
@@ -0,0 +1,16 @@
+from django.contrib.auth.models import AnonymousUser
+
+
+class MFLPermission(permissions.BasePermission):
```
`MFLPermission` is under-developed.

ngurenyaga commented on setup.py:
```diff
@@ -13,5 +13,26 @@
     packages=find_packages(),
-    install_requires=[],
+    install_requires=[
+        "django==1.8.0",
```
Use the "greater than but less than" style to constrain dependencies.

This style will keep us from getting security patches ( or require manual code changes ). 

Apply this principle - with discretion - on all the dependencies.

ngurenyaga commented on setup.py:
```diff
@@ -13,5 +13,26 @@
+        "markdown==2.5.1",
+        "python-dateutil==2.4.0",
+        "flake8==2.3.0",
+        "django-grappelli==2.6.3",
```
Remove `grapelli` - both here and in settings ( if applicable ).

We are completely getting rid of the admin "crutch".

ngurenyaga commented on users/models.py:
```diff
@@ -0,0 +1,97 @@
+    is_staff = models.BooleanField(default=False)
+    is_active = models.BooleanField(default=True)
+    date_joined = models.DateTimeField(default=timezone.now)
+    is_incharge = models.BooleanField(default=False)
```
I'd remove the `is_incharge` field and just leave the association with the county.

Then - in custom permissions - I'd express the "level" of access ( in charge, data entry only, review only etc ).

This field will lead to some very intractable permissions logic.

Naturally, this removal will have cascading effects in the custom permissions ( one good reason to centralize the implementation in one place ).

ngurenyaga commented on users/models.py:
```diff
@@ -0,0 +1,97 @@
+        return user
+
+
+class MflUser(AbstractBaseUser, PermissionsMixin):
```
Class naming conventions...where "MFL" is an abbreviation, the class should be named "MFLUser".

ngurenyaga commented on users/models.py:
```diff
@@ -0,0 +1,97 @@
+        super(MflUser, self).save(*args, **kwargs)
+
+
+class UserInchargeCounties(AbstractBase):
```
Good try...

Another rename - to `CountyUser`

ngurenyaga commented on users/models.py:
```diff
@@ -0,0 +1,97 @@
+    """
+    Will store a record of the counties that a user has been incharge of
+    """
+    user = models.ForeignKey(MflUser, related_name='counties')
```
That related name could be more expressive...e.g `user_counties`.

ngurenyaga commented on users/models.py:
```diff
@@ -0,0 +1,97 @@
+
+    def current_active_counties(self):
+        """
+        A user can be incharge of several counties at the same time.
```
I hope you realize the dramatic implications of that sentence...you'd have to implement a chooser in the UI and sophisticated filtering logic is serializers so that you can enforce the invariant that "a county level user works on only one county at a time".

I'd much prefer that you remove this _then_ add a validator that enforces the following rule: "a user can only be active in one county at a time".

ngurenyaga commented on users/views.py:
```diff
@@ -0,0 +1,71 @@
+class UserDetailView(FilterViewMixin, generics.RetrieveUpdateDestroyAPIView):
+    queryset = MflUser.objects.all()
+    serializer_class = UserSerializer
+    lookup_field = 'id'
```
`lookup_field` does not change default behavior.

ngurenyaga commented on users/views.py:
```diff
@@ -0,0 +1,71 @@
+class UserList(FilterViewMixin, generics.ListCreateAPIView):
+    queryset = MflUser.objects.all()
+    serializer_class = UserSerializer
+    filter_fields = ('is_active', 'is_incharge', 'is_superuser', 'is_staff')
```
Filter classes...already stated that it _appears_ to be overkill at present but is actually not.

ngurenyaga commented on users/views.py:
```diff
@@ -0,0 +1,71 @@
+class UserCountiesView(FilterViewMixin, generics.ListCreateAPIView):
+    queryset = InchargeCounties.objects.all()
+    serializer_class = InchargeCountiesSerializer
+    filter_fields = ('user', 'county', )
```
Filter classes

ngurenyaga commented on users/views.py:
```diff
@@ -0,0 +1,71 @@
+        FilterViewMixin, generics.RetrieveUpdateDestroyAPIView):
+    queryset = InchargeCounties.objects.all()
+    serializer_class = InchargeCountiesSerializer
+    lookup_field = 'id'
```
This `lookup_field` does not change default behavior.

ngurenyaga commented on facilities/models.py:
```diff
@@ -0,0 +1,83 @@
+    name = models.CharField(max_length=100)
+
+
+class Facility(AbstractBase):
```
You have not provided a way to link a facility to its contacts.

ngurenyaga commented:
# Second Review Checklist, Part One

@bmarika I have unchecked ten boxes in your first review ( the items described there were not done or not done completely - see detailed comments for more detail ). Before you tick a checklist box, kindly make certain that it is actually completed.

This is "Part One" of the second review because the review of the specifications and input documents is still ongoing.

Also - kindly make one commit per check-list item and in the commit description copy some of the check-list item text; that will make it quick and easy for me to review.
- [x] Improve the docstrings; see comments on specific docstrings that are deficient, then review to ensure that each model has appropriate docstrings. **The natural source of information for docstrings and `help_text` is the implementation guide**
- [x] Link the `Facility` model to the refactored `Contact` model ( missing M2M )
- [x] DRYness in the `fabfile`; half done, complete ( see detailed comments )
- [x] Callable to generate sequential and unique facility codes; set up as the default for the `Facility` model's `code` field. In the doc-string, explain how you plan to migrate existing facility codes ( if your sequential codes will have a numeric offset, there should be an SQL migration that effects that ). After doing this, you will remove the custom `save` method ( unless it has an extra purpose ). In order to allow for porting of existing codes, our new code sequence for facilities shall start from 100001. You could also choose to make this the "signature" of MFL v2 by making all our sequences start from there.
- [x] Callable to generate sequential and unique `Owner` codes ( as above )
- [x] Callable to generate sequential and unique `Service` codes ( as above )
- [x] Callable to generate sequential `RegionAbstractBase` codes ( as above )
- [x] Renaming of the facility GIS model - see comments; plus change the way in which it links to facilities; it is a facility that has GIS coordinates, not GIS coordinates that have a facility
- [x] Filter classes for **all** list endpoints; non-negotiable 
- [x] Ordering for **all** list endpoints, set up `OrderingFilter` in settings; non-negotiable
- [x] Rename the model that links users to counties and implement refactoring that is proposed in the comments
- [x] Multiple grammar and content fixes in the `README` - see the comments
- [x] Include a section on authentication in the `README`; take this section to consider the divergent needs of front-end developers and pure API consumers ( one group would be happy with OAuth or even tokens; another group happy with Session - do you support both or do you bias to one side? Why? )
- [x] Implement API root and amend `README` accordingly
- [x] Document content negotiation via headers in `README`; it is customary to use example `curl` commands for that purpose
- [x] Model to configure contact types ( no choice field ); default data for it
- [x] Pervasive `help_text` for model fields; good enough to be rendered by the front-ends. Remember - the threshold for this project in terms of developers and users is lower; take this chance to also fix the minor issues with spaces / newlines
- [x] Default data; esp. for the "setup" models and the models that are replacing choice fields. Port it in via script from the existing MFL - a chance to also find compatibility bottlenecks between the two versions
- [x] **UUID Primary Keys**
- [x] Rewrite the `FilterViewMixin` in line with the comments made
- [x] Get rid of the superflous `lookup_field`s
- [x] Sensible ( hand crafted ) `__unicode__` methods through and through
- [x] Decide if we are using the `HyperlinkedModelSerializer` or the `ModelSerializer` as our default, and why ( explain in a one or two line comment )
- [x] The data bootstrap management command should use `self.stdout.write`
- [x] Make `FacilityStatus` and `FacilityType` names unique
- [x] Rethink the `open_whole_day` and `open_whole_week` booleans in the `FacilityModel` ( see detailed comment )
- [x] Remove the `mfl.sql` dump
- [x] Seriously rethink the place of the `roles` app
- [x] `MFLPermission` is underdeveloped...either work on it or add a clearly worded ticket to defer that work
- [x] Fix the requirements in `setup.py` ( see detailed comments )
- [x] Remove the `is_incharge` boolean from the custom user model and sort out its cascading effects; also make the slight adjustment to the model name that is proposed 
- [x] Rename the `InChargeCounties` model and fix the `related_name` in the `user` model. 
- [x] Add the validator that a user can only be active in one county at a time
- [ ] Complete test coverage

Commit:
Tests checkpoint.

Commit:
Enhance DRYness in the fabfile.

Commit:
Rename GIS model to FacilityGIS
Also added the facillity field as a OneToOne to facility

Commit:
Fix typos.

Commit:
Make id a uuid field.
Add contact_type model to store contact types

Commit:
Add facility contact model.
Add facilty service model to avoid posting many to many objects in views.

Commit:
Replace print with stdout.write.

Commit:
Remove sql file in code.

Commit:
Add unicodes through and through.

Commit:
Remove default serializerc classes and add a common on that matter,

Commit:
Rename userIncharge counties model to usercounties.
Change MFLUser related_name in UserCounties to user_counties.

Commit:
Add a validator in user counties to ensure that user can only be active in one county at a time.

Commit:
Remove Filterable mixin

Commit:
Checkpoint: add docstrings and help text to models.

ngurenyaga commented:
@bmarika The position on "filter classes for everything" has just been reverted; every list endpoint will still need a **filter class**.

The list endpoints will be adapted for reporting uses too; just take my judgement on that - it will eventually be easier that way.

Commit:
Add doctrings and help text to models.

Commit:
Add filter classes for the list endpoints.

Commit:
Checkpoint.

ngurenyaga commented:
# Second Review Checklist, part two

[ this is a living document; it will be get constant updates for the duration of the source material review ]
- [x] Add an `active` field to `AbstractBase`, with the default set to `True`. This field should feature in all filter classes ( use a mixin or shared base if need be ). In the current data model, there is a similar field in many of the tables.
- [x] Add a job titles model; it will be used to organize data about individuals in charge of facilities. The current data model has an equivalent.
- [x] Add `FacilityTypeCategories` model, and the necessary FK in `FacilityType` ( and serializers, views, tests )
- [x] Add support for facility classification ( *\* consult about this one; short version - facilities may be classified into KEPH levels 1 - 6 e.g a Level 6 is a referral hospital ). But it is not that simple - a facility may have departments within it e.g labs and pharmacies that are regulated separately and have separate classifications. See page 79 ( it is some kind of checklist / form ) of the implementation guide.
- [x] Review each model that has join tables e.g `Facility` and confirm that M2M fields are set up for each applicable join table. For example, `Facility` is missing M2M fields for `services`, `contacts` etc.
- [x] Facility `Owner`s will also have categories e.g "Christian Health Association of Kenya", "Kenya Episcopal Conference - Catholic Secretariat" and "Supreme Council for Kenya Muslims" are owners under the category "Faith Based Organization"
- [x] Consider how to model departments / services within a facility that may have separate regulation e.g a lab within a pharmacy. **If you cannot tackle this right away, make a ticket**.
- [x] Consider how to link facilities etc to licensed practitioners; this will be at the heart of the regulator interface and is **important**
- [x] `Service` catalog; services also have categories - `ServiceCategory`
- [ ] Consider how to model the "range of possible values" for a `Service` - the current implementation has a "possible values" field whose choices are "Yes/No", "1-6" or "B-C". **That is not good enough**. Consult the MFL Implementation Guide appendices ( facility data forms )
- [ ] Facility ratings as currently implemented have the following fields: rating of services, rating of cleanliness, rating of attitude, whether the user will return or not, rating date ( obvious ) and occupation of the rater. We do not need to follow this slavishly; but it is a good starting point
- [x] Remove sub-counties as an administrative area and replace with wards ( there are 1450 wards in total, at present )
- [x] Add `editable=False` to all the `code` fields; that hint will be relied upon ( via metadata ) by front-ends

Commit:
minor changes

Commit:
Tests for common app

Commit:
Add users app tests.

Commit:
Checkpoint.

ngurenyaga commented:
@bmarika I'd have added this myself but it has a dependency on work that you are doing in this branch:
- [x] Add `-T template_postgis` to the `fab setup`'s database creation command e.g `createdb -T template_postgis geodjango`
- [x] Add to the `README` a section on system dependencies. The following are system dependencies at present: `postgis`, `geos` and `proj` and `gdal` - installed on Ubuntu with `sudo apt-get install binutils postgis gdal-bin libproj-dev libgeoip1` and `sudo apt-get build-dep python-shapely python-gdal python-numpy cython`
- [x] Add `shapely>=1.5.7` to dependencies
- [x] Add to the `README` a note that "because we use `GDAL`, which is not threadsafe, this server should not be deployed behind a WSGI server that relies on green threads or OS threads'.
- [x] Add the following test / dev dependencies: `fabric pyyaml tox coverage pytest pytest-cov wheel sphinx sphinxcontrib-napoleon sphinx_rtd_theme`. For this project, do not make a distinction between them and "runtime dependencies". Set things up so that a `pip install -e .` from the project checkout folder in a fresh virtualenv gets everything needed to develop or run

Commit:
Added facility tests.

Commit:
Added migrations

ngurenyaga committed:
Added license and classifier information to setup.py

First step towards getting a publishable library

ngurenyaga committed:
Added test setup to setup.py

ngurenyaga committed:
Added a more developed .gitignore

ngurenyaga committed:
Removed test settings file

Will now no longer be relevant - we have a hard dependency on PostGIS

ngurenyaga committed:
Unified version creation into one place

ngurenyaga committed:
Linked setup.py to tox

ngurenyaga committed:
More setup fixes

Commit:
Fix failing tests.

Commit:
Fix model_mommy issues with django==1.8

Commit:
Lock model_mommy to 1.2.4

ngurenyaga committed:
Improvements to tox.ini

ngurenyaga committed:
Made settings more secure by default

This closes #59

The following setting has not been resolved, intentionally.

```
WARNINGS:
?: (security.W004) You have not set a value for the SECURE_HSTS_SECONDS setting. If your entire site is served only over SSL, you may want to consider setting a value and enabling HTTP Strict Transport Security. Be sure to read the documentation first; enabling HSTS carelessly can cause serious, irreversible problems.
```

ngurenyaga committed:
Made settings more compact

ngurenyaga committed:
Added a DEPLOYING.rst file

Commit:
Added API root.

ngurenyaga committed:
Fabfile fixes

Commit:
Add Swagger for the api exploration.

Commit:
Checkpoint updating the readme

Commit:
Merge branch 'master' of github.com:MasterFacilityList/mfl_api into common_models

Conflicts:
	config/settings/base.py

Commit:
Update Read me

Commit:
Minor

ngurenyaga committed:
Merge branch 'common_models' into enhance_setup_py_publish_to_pip

Conflicts:
	.gitignore
	config/__init__.py
	config/settings/base.py
	fabfile.py
	setup.py

ngurenyaga committed:
Removed dangling print statement

ngurenyaga committed:
Added pytest.ini file

Commit:
Create custom sequences fot generating codes.

Commit:
Fix issues in sequencies

ngurenyaga committed:
Fixed test failures

Also altered CircleCI settings

ngurenyaga committed:
Minor change to pytest config

Will now collect only in local dir

ngurenyaga committed:
Minor fix

More CircleCI wrangling

ngurenyaga committed:
Added norecursedirs to tox.ini

ngurenyaga committed:
Added venv to pytest.ini norecursedirs

ngurenyaga committed:
Added build/ to norecursedirs in pytest.ini

One more source of test collection snafus removed

ngurenyaga committed:
Fixed stigmata of merge conflict

bmarika committed:
Merge pull request #73 from MasterFacilityList/enhance_setup_py_publish_to_pip

Enhance setup py publish to pip

Commit:
Add tests to check the generation of codes by sequences at the database level.

ngurenyaga committed:
Added on_delete=models.PROTECT to ForeignKeys

The default cascading delete behavior can be very damaging.
This change makes life a little harder for "genuine" deletes,

Closes #49

Commit:
Merge branch 'common_models' of github.com:MasterFacilityList/mfl_api into common_models

Conflicts:
	config/settings/test.py
	setup.py

ngurenyaga committed:
Merge branch 'common_models' into disable_cascading_deletes

bmarika committed:
Merge pull request #74 from MasterFacilityList/disable_cascading_deletes

Added on_delete=models.PROTECT to ForeignKeys

ngurenyaga committed:
Moved is_classified boolean from the FacilityGPS to Facility model

It is a property of the facility

ngurenyaga committed:
Migration after moving is_classified from the FacilityGPS to the Facility model

ngurenyaga committed:
Added is_classified to the Facility filter class

This will be an important filter for the front end

bmarika committed:
Merge pull request #75 from MasterFacilityList/classified_facilities

Classified facilities

Commit:
Add an active field in abstract base.

Commit:
Merge branch 'common_models' of github.com:MasterFacilityList/mfl_api into common_models

Conflicts:
	facilities/models.py

Commit:
Fix failing migrations.

ngurenyaga committed:
Checkpoint - using auth model FKs

ngurenyaga committed:
Checkpoint

Commit:
Chekpoint

Commit:
Rename sub counties to ward.

ngurenyaga committed:
Checkpoint - successful fab setup after changes to AbstractBase

ngurenyaga committed:
Re-established passing of tests after changes to abstract base

ngurenyaga committed:
Merge commit 'cb8778323497fc6b94dc5553b442ce2843733b50' into disable_deleting

Conflicts:
	common/models.py
	common/tests/test_views.py
	facilities/models.py

ngurenyaga committed:
Merge branch 'common_models' into disable_deleting

Conflicts:
	common/migrations/0001_initial.py
	common/models.py
	common/tests/test_models.py
	common/tests/test_views.py
	facilities/migrations/0001_initial.py
	users/migrations/0001_initial.py

bmarika committed:
Merge pull request #80 from MasterFacilityList/disable_deleting

Disable deleting

ngurenyaga committed:
Moved all non user stuff out of the users app

This slimming down will make it less likely to be a source of
cyclic dependency issues.

This closes #81

ngurenyaga committed:
First ( experimental ) working date filter

Not a mixin yet, no tests

ngurenyaga committed:
Checkpoint - existing tests pass

bmarika committed:
Merge pull request #82 from MasterFacilityList/slim_down_users_app

Moved all non user stuff out of the users app

ngurenyaga committed:
Re-established test coverage

ngurenyaga committed:
Merge branch 'common_models' into mixin_to_filter_all_lists_by_dates

Conflicts:
	common/filters.py
	common/views.py

ngurenyaga committed:
Test with simple filter passing again

ngurenyaga committed:
Test for is_deleted filter passed

ngurenyaga committed:
Now excluding passwords from user serialization

Fixes #64

ngurenyaga committed:
Added common filter base to filters from the facilities app

Commit:
Checkpoint: Before merging in new filters

bmarika committed:
Merge pull request #84 from MasterFacilityList/mixin_to_filter_all_lists_by_dates

Mixin to filter all lists by dates [ work in progress ]

Commit:
Chekpoint.

Commit:
Merge remote-tracking branch 'origin/common_models' into add_test_for_facility_model

Conflicts:
	facilities/filters.py
	facilities/migrations/0001_initial.py
	facilities/models.py

ngurenyaga committed:
Added collection date to facility GIS model

This closes #60

ngurenyaga committed:
Merge pull request #86 from MasterFacilityList/toughen_tests

Toughen tests

ngurenyaga committed:
Merge branch 'common_models' into multiple_enhancements

Commit:
Fix typo

ngurenyaga committed:
Added django-rest-auth

Not tested yet

ngurenyaga committed:
Added is_published boolean field

When is_published is True, the facility will be visible on the
public site.

Commit:
Add user residence model.

ngurenyaga committed:
Multiple model bug fixes

bmarika committed:
Merge pull request #88 from MasterFacilityList/multiple_enhancements

Multiple enhancements

Commit:
Add user contact model

Commit:
Merge branch 'common_models' of github.com:MasterFacilityList/mfl_api into user_model_enhancement_issue_no_31

Commit:
Add six to requirements.

ngurenyaga committed:
Merge pull request #87 from MasterFacilityList/user_model_enhancement_issue_no_31

User model enhancement issue no 31

ngurenyaga committed:
First implementation of SequenceField

Will try to streamline it by removing the custom save methods

ngurenyaga committed:
Tests with new sequence field passing

Coverage misses

ngurenyaga committed:
Fixed missed branch coverage

bmarika committed:
Merge pull request #89 from MasterFacilityList/proper_sequence_field

Proper sequence field

Commit:
Add a facility units model.

Commit:
Merge remote-tracking branch 'origin/common_models' into facility_units_model

Commit:
Add physical address field.

Commit:
Add views for facility units.

ngurenyaga committed:
Merge pull request #90 from MasterFacilityList/facility_units_model

Facility units model

ngurenyaga committed:
Added Town model

bmarika committed:
Merge pull request #91 from MasterFacilityList/fix_town_and_address

Added Town model

ngurenyaga committed:
Installed django-reversion

ngurenyaga committed:
Minor refactor - created models package in the common app

ngurenyaga committed:
Registered all models with reversion

ngurenyaga committed:
Created serializers package in the common app

This will allow the code to be more organized

ngurenyaga committed:
Factored out base serializer mixin

ngurenyaga committed:
Merge pull request #92 from MasterFacilityList/add_django_reversion

Add django reversion

ngurenyaga committed:
Added creation of PostGis template to fab setup

This prepares the system for GIS work

ngurenyaga committed:
Added README note on system dependencies

ngurenyaga committed:
Added thread safety note to README

ngurenyaga committed:
Added docs dependencies

ngurenyaga committed:
Fix dependency issue with sphinx

ngurenyaga commented:
@bmarika I have added some commits to sort out some of the smaller checkboxes.

I have moved some of the pending stuff out to issues in the issue list. The current issues have checklists on them - do not close the issues, I will close them after I verify that every checklist item has been sorted out.

I will merge this PR, so that we can establish a proper `develop` branch and a proper deploy workflow.

ngurenyaga committed:
Removed sphinx
